File: packages/gbnf-experiment/data/20260909T190508Z_7c1544c0/ported_implementation/gbnf/utils/errors/build_error_position.py
from typing import List

MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3


def build_error_position(src: str, pos: int) -> List[str]:
    if src == '':
        return ['No input provided']

    lines = src.split('\n')

    line_idx = 0
    while line_idx < len(lines) and lines[line_idx] and pos > len(lines[line_idx]) - 1:
        pos -= len(lines[line_idx])
        line_idx += 1

    lines_to_show: List[str] = []
    for i in range(max(0, line_idx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1)), line_idx + 1):
        # the line index can run past the end of the source, in which case the
        # reference implementation renders an empty line.
        lines_to_show.append(lines[i] if i < len(lines) else '')

    return [
        *lines_to_show,
        ' ' * pos + '^',
    ]

File: proposed-projects/round-trip-experiment/reverse/20260909T190508Z_7c1544c0/20260910T103012Z_7e3d5183/ported_implementation/src/utils/errors/build-error-position.ts
import { length } from '../js.js';

export const MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3;

export const buildErrorPosition = (src: string, pos: number): string[] => {
  if (src === '') {
    return ['No input provided'];
  }

  const lines = src.split('\n');

  let lineIdx = 0;
  while (
    lineIdx < lines.length &&
    lines[lineIdx] !== '' &&
    pos > length(lines[lineIdx]) - 1
  ) {
    pos -= length(lines[lineIdx]);
    lineIdx += 1;
  }

  const linesToShow: string[] = [];
  for (
    let i = Math.max(0, lineIdx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1));
    i <= lineIdx;
    i++
  ) {
    // the line index can run past the end of the source, in which case the
    // reference implementation renders an empty line.
    linesToShow.push(i < lines.length ? lines[i] : '');
  }

  return [...linesToShow, `${' '.repeat(Math.max(pos, 0))}^`];
};
